### PR TITLE
Added optional Keyvault Name at High Level Data Setup

### PR DESCRIPTION
**Pull request by adetolacog**
Notes:
* Product key vault name was taken for fis, so keyvault was created with custom name other than product name.

### CODE REVIEW

**Review by timja — CHANGES_REQUESTED** [review]
see https://hmcts-reform.slack.com/archives/C8SR5CAMU/p1664805297521929?thread_ts=1664292696.377869&cid=C8SR5CAMU

**Review by timja — COMMENTED** [review]
Please add a proper PR title and description

File: vars/withDefinitionImportSecretsAndEnvVars.groovy [hunk 1]
```diff
@@ -1,5 +1,5 @@
 
-def call(String product, String environment, Map<String, String> vaultOverrides, Closure body) {
+def call(String product, String environment, String highLevelDataSetupKeyVaultName, Map<String, String> vaultOverrides, Closure body) {
```
Comment (timja) [inline vars/withDefinitionImportSecretsAndEnvVars.groovy:2]: please refactor this to just pass the vault name in as the product, then everything should work.

File: vars/withDefinitionImportSecretsAndEnvVars.groovy [hunk 2]
```diff
@@ -34,13 +34,19 @@ def call(String product, String environment, Map<String, String> vaultOverrides,
   }
 }
 
-def executeClosure(Iterator<Map.Entry<String,List<Map<String,Object>>>> secretIterator, String product, String dependedEnv, Closure body) {
+def executeClosure(Iterator<Map.Entry<String,List<Map<String,Object>>>> secretIterator, String product, String dependedEnv, String highLevelDataSetupKeyVaultName, Closure body) {
```
Comment (timja) [inline vars/withDefinitionImportSecretsAndEnvVars.groovy:37]: revert the changes here

File: vars/withDefinitionImportSecretsAndEnvVars.groovy [hunk 3]
```diff
@@ -34,13 +34,19 @@ def call(String product, String environment, Map<String, String> vaultOverrides,
...
   // this assumes that the team, a) has a product vault, b) is named the same as their product
   def productName = entry.key != '${product}' ? entry.key : product
-  String theKeyVaultUrl = "https://${productName}-${dependedEnv}.vault.azure.net/"
+    String theKeyVaultUrl = ""
+    if (!highLevelDataSetupKeyVaultName?.trim()) {
+        theKeyVaultUrl = "https://${productName}-${dependedEnv}.vault.azure.net/"
+    }
+    else {
+        theKeyVaultUrl = "https://${highLevelDataSetupKeyVaultName}-${dependedEnv}.vault.azure.net/"
+    }
```
Comment (timja) [inline vars/withDefinitionImportSecretsAndEnvVars.groovy:49]: revert changes here

File: vars/highLevelDataSetup.groovy [hunk 4]
```diff
@@ -13,7 +13,7 @@ def call(params) {
 
   if (config.highLevelDataSetup) {
     stageWithAgent("High Level Data Setup - ${environment}", product) {
-      withDefinitionImportSecretsAndEnvVars(product, environment, config.vaultEnvironmentOverrides){
+      withDefinitionImportSecretsAndEnvVars(product, environment, config.highLevelDataSetupKeyVaultName, config.vaultEnvironmentOverrides){
```
Comment (timja) [inline vars/highLevelDataSetup.groovy:16]: this is where you should check if the config value is non null and instead send that name instead of the product

File: vars/highLevelDataSetup.groovy [hunk 5]
```diff
@@ -12,8 +12,12 @@ def call(params) {
   def product = params.product
 
   if (config.highLevelDataSetup) {
+    def highLevelDataSetupKeyVaultName = config.highLevelDataSetupKeyVaultName
+
     stageWithAgent("High Level Data Setup - ${environment}", product) {
-      withDefinitionImportSecretsAndEnvVars(product, environment, config.vaultEnvironmentOverrides){
+        def productName = !highLevelDataSetupKeyVaultName?.trim() ? product : highLevelDataSetupKeyVaultName
+
+        withDefinitionImportSecretsAndEnvVars(productName, environment, config.vaultEnvironmentOverrides){
```
Comment (timja) [inline vars/highLevelDataSetup.groovy:20]:
```suggestion
        def vaultName = !highLevelDataSetupKeyVaultName?.trim() ? product : highLevelDataSetupKeyVaultName

        withDefinitionImportSecretsAndEnvVars(vaultName, environment, config.vaultEnvironmentOverrides){
```

**Review by timja — CHANGES_REQUESTED** [review]
Please address
https://github.com/hmcts/cnp-jenkins-library/pull/904#pullrequestreview-1128438420

and can you look at the build failure please

File: vars/highLevelDataSetup.groovy (hunk 4 above)
Comment (adetolacog) [reply vars/highLevelDataSetup.groovy]: Rework completed

File: vars/withDefinitionImportSecretsAndEnvVars.groovy (hunk 3 above)
Comment (adetolacog) [reply vars/withDefinitionImportSecretsAndEnvVars.groovy]: Changes reverted

**Review by timja — APPROVED** [review]
code looks good, just needs a green build sorted.

would be nice to have added reasoning to the PR description as well.
From what I understand your prod vault name was already taken so you had to rename your vault, that would've been good background here.

**Review by adusumillipraveen — APPROVED** [review]
